File: urbanairship-core/src/main/java/com/urbanairship/messagecenter/MessageWebView.java
package com.urbanairship.messagecenter;

import android.annotation.SuppressLint;
import android.content.Context;
import android.util.AttributeSet;

import com.urbanairship.Logger;
import com.urbanairship.UAirship;
import com.urbanairship.richpush.RichPushMessage;
import com.urbanairship.richpush.RichPushUser;
import com.urbanairship.webkit.AirshipWebView;

import java.util.HashMap;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

/**
 * A web view that sets settings appropriate for Airship message center content.
 */
public class MessageWebView extends AirshipWebView {

    private RichPushMessage currentMessage;

    /**
     * MessageWebView Constructor
     *
     * @param context A Context object used to access application assets.
     */
    public MessageWebView(@NonNull Context context) {
        super(context);
    }

    /**
     * MessageWebView Constructor
     *
     * @param context A Context object used to access application assets.
     * @param attrs An AttributeSet passed to our parent.
     */
    public MessageWebView(@NonNull Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);
    }

    /**
     * MessageWebView Constructor
     *
     * @param context A Context object used to access application assets.
     * @param attrs An AttributeSet passed to our parent.
     * @param defStyle The default style resource ID.
     */
    public MessageWebView(@NonNull Context context, @Nullable AttributeSet attrs, int defStyle) {
        super(context, attrs, defStyle);
    }

    /**
     * MessageWebView Constructor
     *
     * @param context A Context object used to access application assets.
     * @param attrs An AttributeSet passed to our parent.
     * @param defStyle The default style resource ID.
     * @param defResStyle A resource identifier of a style resource that supplies default values for
     * the view, used only if defStyle is 0 or cannot be found in the theme. Can be 0 to not
     * look for defaults.
     */
    public MessageWebView(@NonNull Context context, @Nullable AttributeSet attrs, int defStyle, int defResStyle) {
        super(context, attrs, defStyle, defResStyle);
    }

    /**
     * Loads the web view with the rich push message.
     *
     * @param message The RichPushMessage that will be displayed.
     */
    @SuppressLint("NewApi")
    public void loadRichPushMessage(@Nullable RichPushMessage message) {
        if (message == null) {
            Logger.error("Unable to load null message into MessageWebView");
            return;
        }

        RichPushUser user = UAirship.shared().getInbox().getUser();

        // Send authorization in the headers if the web view supports it
        HashMap<String, String> headers = new HashMap<>();

        // Set the auth
        if (user.getId() != null && user.getPassword() != null) {
            setClientAuthRequest(message.getMessageBodyUrl(), user.getId(), user.getPassword());
            headers.put("Authorization", createBasicAuth(user.getId(), user.getPassword()));
        }

        loadUrl(message.getMessageBodyUrl(), headers);

        // loadUrl clears currentMessage, so set it after load starts
        currentMessage = message;
    }

    /**
     * The current loaded RichPushMessage.
     *
     * @return The current RichPushMessage that was loaded.
     */
    @Nullable
    public RichPushMessage getCurrentMessage() {
        return currentMessage;
    }

    /**
     * Called right before data or a URL is passed to the web view to be loaded.
     */
    @SuppressLint("NewApi")
    protected void onPreLoad() {
        currentMessage = null;

        super.onPreLoad();
    }


}

File: urbanairship-message-center/src/main/java/com/urbanairship/messagecenter/ui/widget/MessageWebView.kt
package com.urbanairship.messagecenter.ui.widget

import android.content.Context
import android.os.Bundle
import android.util.AttributeSet
import android.webkit.WebView
import androidx.annotation.CallSuper
import androidx.annotation.MainThread
import androidx.annotation.RestrictTo
import com.urbanairship.UALog
import com.urbanairship.actions.ActionArguments
import com.urbanairship.actions.ActionRunRequest
import com.urbanairship.javascript.JavaScriptEnvironment
import com.urbanairship.json.JsonMap
import com.urbanairship.json.JsonValue
import com.urbanairship.messagecenter.Message
import com.urbanairship.messagecenter.MessageCenter
import com.urbanairship.webkit.AirshipWebView
import com.urbanairship.webkit.AirshipWebViewClient
import java.text.SimpleDateFormat
import java.util.Locale
import java.util.TimeZone
import kotlinx.coroutines.runBlocking

/** Base WebView configured for Airship Message Center content. */
internal class MessageWebView @JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet? = null,
    defStyle: Int = 0,
    defResStyle: Int = 0
): AirshipWebView(context, attrs, defStyle, defResStyle) {

    /**
     * Loads the web view with the [Message].
     *
     * @param message The message that will be displayed.
     */
    fun loadMessage(message: Message) {
        UALog.v { "Loading message: ${message.messageId}" }
        val user = MessageCenter.shared().user

        // Send authorization in the headers if the web view supports it
        val headers = HashMap<String, String>()

        // Set the auth
        val (userId, password) = user.id to user.password
        if (userId != null && password != null) {
            setClientAuthRequest(message.messageBodyUrl, userId, password)
            headers["Authorization"] = createBasicAuth(userId, password)
        }
        UALog.v { "Load URL: ${message.messageBodyUrl}" }
        loadUrl(message.messageBodyUrl, headers)
    }
}

/** A `WebViewClient` that enables the Airship Native Bridge for Message Center. */
internal open class MessageWebViewClient : AirshipWebViewClient() {

    /**
     * @hide
     */
    @RestrictTo(RestrictTo.Scope.LIBRARY_GROUP)
    override fun extendActionRequest(
        request: ActionRunRequest,
        webView: WebView
    ): ActionRunRequest {
        val metadata = Bundle()
        val message = getMessage(webView)
        if (message != null) {
            metadata.putString(ActionArguments.RICH_PUSH_ID_METADATA, message.messageId)
        }
        request.setMetadata(metadata)
        return request
    }

    /**
     * @hide
     */
    @CallSuper
    @RestrictTo(RestrictTo.Scope.LIBRARY_GROUP)
    override fun extendJavascriptEnvironment(
        builder: JavaScriptEnvironment.Builder,
        webView: WebView
    ): JavaScriptEnvironment.Builder {
        val message = getMessage(webView)
        var extras = JsonMap.EMPTY_MAP
        if (message != null) {
            extras = JsonValue.wrapOpt(message.extrasMap).optMap()
        }
        return super.extendJavascriptEnvironment(builder, webView)
            .addGetter("getMessageSentDateMS", message?.sentDateMS ?: -1)
            .addGetter("getMessageId", message?.messageId)
            .addGetter("getMessageTitle", message?.title).addGetter(
                "getMessageSentDate",
                if (message != null) DATE_FORMATTER.format(message.sentDate) else null
            ).addGetter("getUserId", MessageCenter.shared().user.id)
            .addGetter("getMessageExtras", extras)
    }

    /**
     * Helper method to get the RichPushMessage from the web view.
     *
     * @param webView The web view.
     * @return The rich push message, or null if the web view does not have an associated message.
     * @note This method should only be called from the main thread.
     */
    @MainThread
    private fun getMessage(webView: WebView): Message? = runBlocking {
        val url = webView.url
        MessageCenter.shared().inbox.getMessageByUrl(url)
    }

    private companion object {
        // TODO(m3-inbox): Use LONG date format and maybe allow a format string to be set via theme attr?
        private val DATE_FORMATTER = SimpleDateFormat("yyyy-MM-dd HH:mm:ss.SSSZ", Locale.US).apply {
            timeZone = TimeZone.getTimeZone("UTC")
        }
    }
}
